File: ci-cd pallavi/backend/middleware/authMiddleware.js
const jwt = require('jsonwebtoken');

/**
 * @desc    Middleware to protect routes and verify JWT
 */
const protect = async (req, res, next) => {
  let token;

  // 1. Check for token in Authorization header (Format: Bearer <token>)
  if (
    req.headers.authorization &&
    req.headers.authorization.startsWith('Bearer')
  ) {
    try {
      // 2. Extract token from header
      token = req.headers.authorization.split(' ')[1];

      // 3. Verify token
      const decoded = jwt.verify(token, process.env.JWT_SECRET || 'fallback_secret');

      // 4. Attach decoded payload (user ID/data) to the request object
      // This allows subsequent middleware/controllers to access the user's info
      req.user = decoded;

      next();
    } catch (error) {
      console.error('JWT Verification Error:', error.message);
      return res.status(401).json({
        success: false,
        message: 'Not authorized, token failed',
      });
    }
  }

  // 5. If no token is found
  if (!token) {
    return res.status(401).json({
      success: false,
      message: 'Not authorized, no token provided',
    });
  }
};

module.exports = { protect };

File: ci-cd pallavi/backend/app/middleware/auth_middleware.py
from functools import wraps
from flask import request, jsonify
import jwt
import os

def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = None
        
        # Check Authorization header
        if 'Authorization' in request.headers:
            auth_header = request.headers['Authorization']
            if auth_header.startswith('Bearer '):
                token = auth_header.split(" ")[1]

        if not token:
            return jsonify({'message': 'Token is missing!', 'success': False}), 401

        try:
            # Decode token
            data = jwt.decode(token, os.getenv('JWT_SECRET', 'default_secret'), algorithms=["HS256"])
            # In a real app, you might fetch the user from DB here
            # current_user = query_db("SELECT * FROM users WHERE id = %s", (data['id'],), one=True)
            request.user = data
        except Exception as e:
            return jsonify({'message': 'Token is invalid!', 'success': False}), 401

        return f(*args, **kwargs)

    return decorated
